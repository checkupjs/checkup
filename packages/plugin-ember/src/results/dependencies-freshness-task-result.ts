import { BaseTaskResult, TaskResult, ui } from '@checkup/core';
import { DepFreshnessInfo } from '../types';

export default class DependenciesFreshnessTaskResult extends BaseTaskResult implements TaskResult {
  depFreshnessInfo!: DepFreshnessInfo;

  stdout() {
    ui.styledHeader(this.meta.taskName);
    ui.blankLine();

    this._writeFreshnessTable(this.depFreshnessInfo);
    ui.blankLine();
    // Todo: should look into color schemes
  }

  json() {
    return {
      meta: this.meta,
      result: {
        depFreshnessInfo: this.depFreshnessInfo,
      },
    };
  }

  pdf() {
    return undefined;
  }

  _writeFreshnessTable(dependencies: DepFreshnessInfo) {
    if (dependencies.tableBody && dependencies.tableBody.length === 0) {
      return;
    }

    ui.table(this._transformToTableData(dependencies.tableBody), {
      package: {},
      current: {},
      wanted: {},
      latest: {},
      packageType: {},
      url: {},
    });
  }

  _transformToTableData(data: Array<String[]>) {
    let result: {
      package: String;
      current: String;
      wanted: String;
      latest: String;
      packageType: String;
      url: String;
    }[] = [];
    // Todo: should look into making it cleaner here
    data.forEach(dependency => {
      const row = {
        package: dependency[0],
        current: dependency[1],
        wanted: dependency[2],
        latest: dependency[3],
        packageType: dependency[4],
        url: dependency[5],
      };
      result.push(row);
    });
    return result;
  }
}
