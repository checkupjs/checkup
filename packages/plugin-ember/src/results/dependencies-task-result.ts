import { BaseTaskResult, Task, TaskResult, toPairs, ui } from '@checkup/core';

import { PackageJson } from 'type-fest';

export default class DependenciesTaskResult extends BaseTaskResult implements TaskResult {
  emberLibraries!: PackageJson.Dependency;
  emberAddons!: Record<string, PackageJson.Dependency>;
  emberCliAddons!: Record<string, PackageJson.Dependency>;

  constructor(task: Task) {
    super(task);
    this.emberLibraries = {};
  }

  get hasDependencies() {
    return (
      Object.keys(this.emberLibraries).length ||
      [this.emberAddons, this.emberCliAddons].every(addons => {
        return Object.keys(addons.dependencies).length && Object.keys(addons.devDependencies);
      })
    );
  }

  stdout() {
    if (!this.hasDependencies) {
      return;
    }
    ui.styledHeader(this.meta.friendlyTaskName);
    this._writeDependencySection('Dependencies - Core Libraries', this.emberLibraries);
    this._writeDependencySection('Dependencies - Ember Addons', this.emberAddons.dependencies);
    this._writeDependencySection(
      'Dev Dependencies - Ember Addons',
      this.emberAddons.devDependencies
    );
    this._writeDependencySection(
      'Dependencies - Ember CLI Addons',
      this.emberCliAddons.dependencies
    );
    this._writeDependencySection(
      'Dev Dependencies - Ember CLI Addons',
      this.emberCliAddons.devDependencies
    );
  }

  json() {
    return {
      meta: this.meta,
      result: {
        emberLibraries: this.emberLibraries,
        emberAddons: this.emberAddons,
        emberCliAddons: this.emberCliAddons,
      },
    };
  }

  pdf() {}

  _writeDependencySection(header: string, dependencies: PackageJson.Dependency) {
    if (Object.keys(dependencies).length === 0) {
      return;
    }

    ui.styledHeader(header);
    ui.blankLine();
    ui.table(toPairs(dependencies, { keyName: 'dependency', valueName: 'version' }), {
      dependency: {},
      version: {},
    });
    ui.blankLine();
  }
}
