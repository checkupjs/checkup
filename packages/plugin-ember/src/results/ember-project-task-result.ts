import { BaseTaskResult, TaskResult, ui } from '@checkup/core';

export default class EmberProjectTaskResult extends BaseTaskResult implements TaskResult {
  type!: string;
  name!: string;
  version!: string;

  toConsole() {
    ui.styledHeader(this.meta.friendlyTaskName);
    ui.blankLine();
    ui.styledObject({
      type: this.type,
    });
    ui.blankLine();
  }

  toJson() {
    return {
      [this.meta.taskName]: { type: this.type },
    };
  }
}
