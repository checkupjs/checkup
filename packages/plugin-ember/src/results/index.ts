export { default as EmberProjectTaskResult } from './ember-project-task-result';
export { default as DependenciesTaskResult } from './dependencies-task-result';
export { default as TypesTaskResult } from './types-task-result';
