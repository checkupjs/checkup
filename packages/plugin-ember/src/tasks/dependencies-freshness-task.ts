import { BaseTask, Category, Priority, Task, TaskMetaData, TaskResult } from '@checkup/core';
import DependenciesFreshnessTaskResult from '../results/dependencies-freshness-task-result';
import { DepFreshnessInfo } from '../types';

const hash = require('promise.hash.helper');
const shell = require('shelljs');
const OUTDATED_DEP = 'yarn outdated --json';

async function getOutdatedDep(): Promise<DepFreshnessInfo> {
  const { stdout } = await shell.exec(OUTDATED_DEP, { silent: true });

  // stripping out color legend info from the output
  const outdatedDepTable = JSON.parse(stdout.split('}')[1] + '}}');
  // todo: error handle when there is no outdatedDep
  return hash({
    tableHead: outdatedDepTable.data.head,
    tableBody: outdatedDepTable.data.body,
  });
}

export default class DependenciesFreshnessTask extends BaseTask implements Task {
  meta: TaskMetaData = {
    taskName: 'dependencies-freshness',
    friendlyTaskName: 'Dependencies Freshness',
    taskClassification: {
      category: Category.Insights,
      priority: Priority.High,
    },
  };

  async run(): Promise<TaskResult> {
    let result: DependenciesFreshnessTaskResult = new DependenciesFreshnessTaskResult(this.meta);
    result.depFreshnessInfo = await getOutdatedDep();

    return result;
  }
}
