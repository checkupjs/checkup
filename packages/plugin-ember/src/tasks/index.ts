export { default as ProjectInfoTask } from './project-info-task';
export { default as DependenciesTask } from './dependencies-task';
export { default as TypesTask } from './types-task';
