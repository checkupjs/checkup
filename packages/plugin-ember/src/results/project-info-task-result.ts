import { TaskResult, ui } from '@checkup/core';

import { ProjectInfoTask } from '../tasks';

export default class ProjectInfoTaskResult implements TaskResult {
  type!: string;
  name!: string;
  version!: string;

  toConsole() {
    ui.styledHeader(ProjectInfoTask.friendlyTaskName);
    ui.blankLine();
    ui.styledObject({
      name: this.name,
      type: this.type,
      version: this.version,
    });
    ui.blankLine();
  }

  toJson() {
    return {
      [ProjectInfoTask.taskName]: { name: this.name, type: this.type, version: this.version },
    };
  }
}
