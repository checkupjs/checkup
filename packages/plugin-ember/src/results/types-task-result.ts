import { TaskItemData, TaskResult, ui } from '@checkup/core';

import { TypesTask } from '../tasks';

export default class TypesTaskResult implements TaskResult {
  types!: TaskItemData[];

  findByType(typeName: string): TaskItemData | undefined {
    return this.types.find(type => type.type === typeName);
  }

  toConsole() {
    ui.styledHeader(TypesTask.friendlyTaskName);
    ui.blankLine();
    ui.table(this.types, { type: {}, total: {} });
    ui.blankLine();
  }

  toJson() {
    return { [TypesTask.taskName]: this.types };
  }
}
