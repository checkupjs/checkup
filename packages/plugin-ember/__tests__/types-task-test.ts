import EmberCLIFixturifyProject from './__utils__/ember-cli-fixturify-project';
import { stdout } from './__utils__/stdout';

import { TypesTask } from '../src/tasks';
import { TypesTaskResult } from '../src/results';

const FILE_PATH = 'tests/test-app';

const TYPES = {
  components: {
    'my-component.js': '',
  },
  controllers: {
    'my-controller.js': '',
  },
  helpers: {
    'my-helper.js': '',
  },
  initializers: {
    'my-initializer.js': '',
  },
  'instance-initializers': {
    'my-helper.js': '',
  },
  mixins: {
    'my-mixin.js': '',
  },
  models: {
    'my-model.js': '',
  },
  routes: {
    'my-route.js': '',
  },
  services: {
    'my-service.js': '',
  },
  templates: {
    'my-component.hbs': '',
  },
};

describe('types-task', () => {
  let fixturifyProject: EmberCLIFixturifyProject;

  beforeEach(function() {
    fixturifyProject = new EmberCLIFixturifyProject('cli-checkup-app', '0.0.0');
  });

  afterEach(function() {
    fixturifyProject.dispose(FILE_PATH);
  });

  it('returns all the types found in the app and outputs to the console', async () => {
    fixturifyProject.files = Object.assign(fixturifyProject.files, {
      'index.js': 'index js file',
      addon: TYPES,
    });

    fixturifyProject.writeSync(FILE_PATH);

    const result = await new TypesTask().run();
    const typesTaskResult = <TypesTaskResult>result;

    typesTaskResult.toConsole();

    expect(stdout()).toMatchSnapshot();
  });

  it('returns all the types (including nested) found in the app and outputs to the console', async () => {
    fixturifyProject.files = Object.assign(fixturifyProject.files, {
      'index.js': 'index js file',
      addon: TYPES,
    });

    fixturifyProject.addInRepoAddon('ember-super-button', 'latest');

    // @ts-ignore
    fixturifyProject.files.lib['ember-super-button'].addon = TYPES;
    // @ts-ignore

    fixturifyProject.writeSync(FILE_PATH);

    const result = await new TypesTask().run();
    const typesTaskResult = <TypesTaskResult>result;

    typesTaskResult.toConsole();

    expect(stdout()).toMatchSnapshot();
  });

  it('returns all the types found in the app and outputs to JSON', async () => {
    fixturifyProject.files = Object.assign(fixturifyProject.files, {
      'index.js': 'index js file',
      addon: TYPES,
    });

    fixturifyProject.writeSync(FILE_PATH);

    const result = await new TypesTask().run();
    const typesTaskResult = <TypesTaskResult>result;

    expect(typesTaskResult.toJson()).toMatchSnapshot();
  });

  it('returns all the types (including nested) found in the app and outputs to JSON', async () => {
    fixturifyProject.files = Object.assign(fixturifyProject.files, {
      'index.js': 'index js file',
      addon: TYPES,
    });

    fixturifyProject.addInRepoAddon('ember-super-button', 'latest');

    // @ts-ignore
    fixturifyProject.files.lib['ember-super-button'].addon = TYPES;
    // @ts-ignore

    fixturifyProject.writeSync(FILE_PATH);

    const result = await new TypesTask().run();
    const typesTaskResult = <TypesTaskResult>result;

    expect(typesTaskResult.toJson()).toMatchSnapshot();
  });
});
