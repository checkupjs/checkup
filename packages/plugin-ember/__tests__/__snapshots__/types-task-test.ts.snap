// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`types-task returns all the types (including nested) found in the app and outputs to JSON 1`] = `
Object {
  "types": Array [
    Object {
      "data": Array [
        "addon/components/my-component.js",
        "lib/ember-super-button/addon/components/my-component.js",
      ],
      "total": 2,
      "type": "components",
    },
    Object {
      "data": Array [
        "addon/controllers/my-controller.js",
        "lib/ember-super-button/addon/controllers/my-controller.js",
      ],
      "total": 2,
      "type": "controllers",
    },
    Object {
      "data": Array [
        "addon/helpers/my-helper.js",
        "lib/ember-super-button/addon/helpers/my-helper.js",
      ],
      "total": 2,
      "type": "helpers",
    },
    Object {
      "data": Array [
        "addon/initializers/my-initializer.js",
        "lib/ember-super-button/addon/initializers/my-initializer.js",
      ],
      "total": 2,
      "type": "initializers",
    },
    Object {
      "data": Array [
        "addon/instance-initializers/my-helper.js",
        "lib/ember-super-button/addon/instance-initializers/my-helper.js",
      ],
      "total": 2,
      "type": "instance-initializers",
    },
    Object {
      "data": Array [
        "addon/mixins/my-mixin.js",
        "lib/ember-super-button/addon/mixins/my-mixin.js",
      ],
      "total": 2,
      "type": "mixins",
    },
    Object {
      "data": Array [
        "addon/models/my-model.js",
        "lib/ember-super-button/addon/models/my-model.js",
      ],
      "total": 2,
      "type": "models",
    },
    Object {
      "data": Array [
        "addon/routes/my-route.js",
        "lib/ember-super-button/addon/routes/my-route.js",
      ],
      "total": 2,
      "type": "routes",
    },
    Object {
      "data": Array [
        "addon/services/my-service.js",
        "lib/ember-super-button/addon/services/my-service.js",
      ],
      "total": 2,
      "type": "services",
    },
    Object {
      "data": Array [
        "addon/templates/my-component.hbs",
        "lib/ember-super-button/addon/templates/my-component.hbs",
      ],
      "total": 2,
      "type": "templates",
    },
  ],
}
`;

exports[`types-task returns all the types (including nested) found in the app and outputs to the console 1`] = `
"=== Project Types

Type                  Total 
components            2     
controllers           2     
helpers               2     
initializers          2     
instance-initializers 2     
mixins                2     
models                2     
routes                2     
services              2     
templates             2     

"
`;

exports[`types-task returns all the types found in the app and outputs to JSON 1`] = `
Object {
  "types": Array [
    Object {
      "data": Array [
        "addon/components/my-component.js",
      ],
      "total": 1,
      "type": "components",
    },
    Object {
      "data": Array [
        "addon/controllers/my-controller.js",
      ],
      "total": 1,
      "type": "controllers",
    },
    Object {
      "data": Array [
        "addon/helpers/my-helper.js",
      ],
      "total": 1,
      "type": "helpers",
    },
    Object {
      "data": Array [
        "addon/initializers/my-initializer.js",
      ],
      "total": 1,
      "type": "initializers",
    },
    Object {
      "data": Array [
        "addon/instance-initializers/my-helper.js",
      ],
      "total": 1,
      "type": "instance-initializers",
    },
    Object {
      "data": Array [
        "addon/mixins/my-mixin.js",
      ],
      "total": 1,
      "type": "mixins",
    },
    Object {
      "data": Array [
        "addon/models/my-model.js",
      ],
      "total": 1,
      "type": "models",
    },
    Object {
      "data": Array [
        "addon/routes/my-route.js",
      ],
      "total": 1,
      "type": "routes",
    },
    Object {
      "data": Array [
        "addon/services/my-service.js",
      ],
      "total": 1,
      "type": "services",
    },
    Object {
      "data": Array [
        "addon/templates/my-component.hbs",
      ],
      "total": 1,
      "type": "templates",
    },
  ],
}
`;

exports[`types-task returns all the types found in the app and outputs to the console 1`] = `
"=== Project Types

Type                  Total 
components            1     
controllers           1     
helpers               1     
initializers          1     
instance-initializers 1     
mixins                1     
models                1     
routes                1     
services              1     
templates             1     

"
`;
